docs/source/developing-volttron/developing-agents/using-asyncio-in-agents.rst

```diff
@@ -0,0 +1,130 @@
+
+* A `Greenlet <https://greenlet.readthedocs.io/en/latest/#instantiation>`_ is a "lightweight coroutine for in-process sequential concurrent programming. Greenlets can be used on their own, but they are frequently used with frameworks such as gevent to provide higher-level abstractions and asynchronous I/O.
+
+* Asyncio is a built-in Python module that allows the developer to write concurrent code.
+
+* Gevent is a "coroutine-based Python networking library that uses greenlet to provide a high-level synchronous API on top of the libev or libuv event loop".
+
+The general steps to use Asyncio within the Volttron Agent framework are the following:
+
+1. Create an async method.
+2. Create a method that instantiates an Asyncio Event Loop, adds the async method created in the previous step to the Event Loop, and then starts the Event Loop.
+3. Create a method that will spawn a Greenlet and then run the Event Loop that was created in the previous step within the Greenlet.
+4. Within the constructor of the agent, use the VOLTTRON™ Interconnect Protocol (VIP) to subscribe to the method created in the previous step.
+
+Below are code examples of how to implement the steps within an agent. For demonstration purposes, we name this agent, ExampleAsyncioAgent.
+
+Step 1: Create an async method.
+
+.. code-block:: python
+
+    class ExampleAsyncioAgent(Agent):
+
+        async def handle_event(self, event):
+            # do things that include a blocking call
+            await asyncio.sleep(1)
+            return "hello!"
+
+
+Step 2. Create a method that instantiates an Asyncio Event Loop, adds the async method created in the previous step to the Event Loop, and then starts the Event Loop.
+
+.. code-block:: python
+
+    class ExampleAsyncioAgent(Agent):
+
+        def _start_asyncio_loop(self):
+            loop = asyncio.get_event_loop()
+            loop.create_task(self.handle_event)
+            loop.run_forever()
+
+
+3. Create a method that will spawn a Greenlet and then run the Event Loop that was created in the previous step within the Greenlet.
+We will name this method 'self._configure_agent'. It contains code to configure our agent including running the Event Loop.
+When Volttron starts the agent, it will call this method. This method uses Gevent to spawn the Greenlet.
+
+.. code-block:: python
+
+    class ExampleAsyncioAgent(Agent):
+
+        def _configure_agent(self, config_name, action, contents):
+            # do some configuration
+            ...
+
+            gevent.spawn_later(3, self._start_asyncio_loop)
+
+
+Step 4. Within the constructor of the agent, use the VOLTTRON™ Interconnect Protocol (VIP) to subscribe to the method, '_configure_agent', created in the previous step.
```

Comment:
Perhaps lets do this in the onstart of the agent rather than in a callback from the config store.  I think that will have greater understanding/relevance.

Reply:
I agree. Will change.